mountpoint-s3/tests/fs.rs

```diff
@@ -164,6 +166,225 @@ async fn test_read_dir_nested(prefix: &str) {
     fs.releasedir(dir_ino, dir_handle, 0).await.unwrap();
 }
 
+#[tokio::test]
+async fn test_lookup_negative_cached() {
+    let fs_config = S3FilesystemConfig {
+        cache_config: CacheConfig {
+            prefer_s3: false,
+            dir_ttl: Duration::from_secs(600),
+            file_ttl: Duration::from_secs(600),
+        },
+        ..Default::default()
+    };
+    let (client, fs) = make_test_filesystem("test_lookup_negative_cached", &Default::default(), fs_config);
+
+    let head_counter = client.new_counter(Operation::HeadObject);
+    let list_counter = client.new_counter(Operation::ListObjectsV2);
+
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect_err("should fail as no object exists");
+    assert_eq!(head_counter.count(), 1);
+    assert_eq!(list_counter.count(), 1);
+
+    // Check no negative caching
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect_err("should fail as no object exists");
+    assert_eq!(head_counter.count(), 2);
+    assert_eq!(list_counter.count(), 2);
+
+    client.add_object("file1.txt", MockObject::constant(0xa1, 15, ETag::for_tests()));
+
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect("should succeed as object exists and no negative cache");
+    assert_eq!(head_counter.count(), 3);
+    assert_eq!(list_counter.count(), 3);
+
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect("should succeed as object is cached and exists");
+    assert_eq!(head_counter.count(), 3);
+    assert_eq!(list_counter.count(), 3);
+}
+
+#[tokio::test]
+async fn test_lookup_then_open_cached() {
+    let fs_config = S3FilesystemConfig {
+        cache_config: CacheConfig {
+            prefer_s3: false,
+            dir_ttl: Duration::from_secs(600),
+            file_ttl: Duration::from_secs(600),
+        },
+        ..Default::default()
+    };
+    let (client, fs) = make_test_filesystem("test_lookup_then_open_cached", &Default::default(), fs_config);
+
+    client.add_object("file1.txt", MockObject::constant(0xa1, 15, ETag::for_tests()));
+
+    let head_counter = client.new_counter(Operation::HeadObject);
+    let list_counter = client.new_counter(Operation::ListObjectsV2);
+
+    let entry = fs.lookup(FUSE_ROOT_INODE, "file1.txt".as_ref()).await.unwrap();
+    let ino = entry.attr.ino;
+    assert_eq!(head_counter.count(), 1);
+    assert_eq!(list_counter.count(), 1);
+
+    let fh = fs.open(ino, S_IFREG as i32, 0).await.unwrap().fh;
+    fs.release(ino, fh, 0, None, true).await.unwrap();
+    assert_eq!(head_counter.count(), 1);
+    assert_eq!(list_counter.count(), 1);
```

Comment:
to the previous comment, this surprised me until I understood what `prefer_s3` meant

Reply:
probably also worth an uncached version of this test that confirms we do the requests again on `open`

Reply:
agree! I'm wondering whether it is enough to find a way to run existing tests with and without caching: if they pass, then it must have gone to S3.

that being said, maybe it's nice to codify the number of requests we expect.

still thinking here.

Reply:
We don't want to go too overboard with hard-coding the exact request counts for all these scenarios in the uncached case, but one uncached test would be helpful.

Reply:
Added a case here for lookup only